Enforce buffer size limits for BlockEncodingBuffer

yingsu00 opened the pull request:
Partially resolves #14162
Depend on https://github.com/prestodb/presto/pull/14286

The byte[] buffers in BlockEncodingBuffer used to be grown by a factor
of 2. This could cause the buffers larger than necessary. Reducing
the expansionFactor has negative impact on performance. Therefore we
want to enforce stricter limit when growing these buffers. This PR
uses incoming blocks' retained sizes as an estimation for how large the
buffers can grow. It shows 30% to 40% memory usage reduction for fixed
width types and 10% reduction for variable width types for the
OptimizedPartitionedOutputOperator.

```
== NO RELEASE NOTE ==
```

mbasmanova commented on presto-array/src/main/java/com/facebook/presto/array/Arrays.java:
```diff
@@ -104,9 +106,9 @@ else if (buffer.length < capacity) {
     }
 
-    public static byte[] ensureCapacity(byte[] buffer, int capacity, ExpansionFactor expansionFactor, ExpansionOption expansionOption, ArrayAllocator allocator)
+    public static byte[] ensureCapacity(byte[] buffer, int capacity, int targetSize, ExpansionFactor expansionFactor, ExpansionOption expansionOption, ArrayAllocator allocator)
```
targetSize -> estimatedMaxCapacity for consistency

yingsu00 replied:
Done. I renamed the BlockEncodingBuffer fields names too. E.g. `valuesBufferTargetSize` --> `estimatedValueBufferMaxCapacity`

mbasmanova commented on presto-array/src/main/java/com/facebook/presto/array/Arrays.java:
```diff
@@ -115,7 +117,7 @@ else if (buffer.length < capacity) {
             newBuffer = allocator.borrowByteArray(newCapacity);
             if (expansionOption == PRESERVE) {
-                System.arraycopy(buffer, 0, newBuffer, 0, Math.min(buffer.length, capacity));
+                System.arraycopy(buffer, 0, newBuffer, 0, min(buffer.length, capacity));
```
min(buffer.length, capacity) -> buffer.length

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/repartition/ArrayBlockEncodingBuffer.java:
```diff
@@ -201,17 +204,19 @@ public String toString()
         populateNestedPositions(columnarArray);
 
-        ((AbstractBlockEncodingBuffer) valuesBuffers).setupDecodedBlockAndMapPositions(decodedBlockNode.getChildren().get(0));
+        ((AbstractBlockEncodingBuffer) valuesBuffers).setupDecodedBlockAndMapPositions(decodedBlockNode.getChildren().get(0), partitionBufferCapacity);
```
Instead of casting in multiple places change the type of valuesBuffers to AbstractBlockEncodingBuffer

yingsu00 replied:
Done in new commit `17ee832a6a Change nested buffers to AbstractBlockEncodingBuffer`

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/repartition/ArrayBlockEncodingBuffer.java:
```diff
@@ -201,17 +204,19 @@ public String toString()
         ColumnarArray columnarArray = (ColumnarArray) decodedBlockNode.getDecodedBlock();
         decodedBlock = columnarArray.getNullCheckBlock();
 
+        setTargetBufferSize(decodedBlockNode, partitionBufferCapacity);
```
I suggest to inline setTargetBufferSize. It is used only once and doesn't actually *set* target size.

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/repartition/DecodedBlockNode.java:
```diff
@@ -42,10 +43,16 @@
+    // The portion of the size of the exclusive content of this block to the total size. For example, the exclusive content
+    // of an ArrayBlock is its offsets and nulls array while not its elementBlock. It is used to estimate the limit of the
+    // buffers' sizes.
+    private double sizePortionInPage;
```
Can this be calculated within this class as getRetainedSizeInBytes - sum(child.getRetainedSizeInBytes())? If not, perhaps, move this change into a separate commit.

yingsu00 replied:
@mbasmanova Yes you're actually right. I moved the logic to DecodedBlockNode and made a new commit `52456a0fec Introduce sizePortionInPage in DecodedBlockNode`

yingsu00 commented:
@mbasmanova Thanks for reviewing! I have addressed all your comments. Will you be able to take another look? Thank you!

mbasmanova commented:
@yingsu00 Thanks. This PR shows conflicts with master. Please, rebase and I'll take another pass.

yingsu00 commented:
> @yingsu00 Thanks. This PR shows conflicts with master. Please, rebase and I'll take another pass.

@mbasmanova Hi Masha, I just rebased. Would you please take a look please?

yingsu00 commented:
@mbasmanova Hi Masha, the motivation is to get an estimate of the buffers' capacity as close as possible. The estimation is based on the retained sizes of the blocks in the current page. For example, a page with two blocks has the following retained sizes:

channel 0: LongArrayBlock(block0) of 100KB
channel 1: ArrayBlock(block1) of 900KB with nested LongArrayBlock(block2) of 800KB
All blocks don't have nulls. Then block0's portion in the page is `100KB/(100KB + 900KB) = 10%`, while block1 's portion (for the offsets) is `(900KB-800KB)/(100KB + 900KB) = 10%`, and block2's portion is `800KB/(100KB + 900KB) = 80%`.

Now suppose the PartitionBuffer's limit is 1MB, then block0's valuesBuffer max capacity estimation is 0.1MB, block1's offsetsBuffer max capacity estimation is 0.1MB, and block2's valuesBuffer max capacity estimation is 0.8MB. If the buffers grow beyond this limit, the expansion factor will become 1.

yingsu00 commented:
@mbasmanova Hi Masha, did my previous reply answer your question? Would adding more comments to the code help? Or do you have other ideas to address the problem? This PR reduces memory consumption by 30% for a lot of cases without impacting the performance too much and I'd love to hear your suggestion and get it merged. Thank you!

mbasmanova commented:
@yingsu00 Ying, thanks for explaining. I'm wondering if this calculation is correct when some blocks are dictionary encoded. Since repartitioning flattens these blocks, the fraction of the original page size for these blocks is less than the fraction of flattened data in the output page. Is this a problem? Would it help to account for that when computing the fractions?

Aside from the question above, I think the code would be simpler if this calculation was localized. Specifically, I suggest the following changes.

- DecodedBlock: Replace sizePortionInPage with childrenRetainedSizeInBytes 

```
        long childrenSize = 0;
        for (int i = 0; i < children.size(); i++) {
            childrenSize += children.get(i).getRetainedSizeInBytes();
        }

        childrenRetainedSizeInBytes = childrenSize;
```

- PagePartitioner: Add pageRetainedSizeInBytes parameter to appendData method and pass this value to setupDecodedBlocksAndPositions

```
            long pageRetainedSizeInBytes = page.getRetainedSizeInBytes();
            for (int i = 0; i < decodedBlocks.length; i++) {
                decodedBlocks[i] = decodeBlock(flattener, blockLeaseCloser, page.getBlock(i), pageRetainedSizeInBytes);
            }

            // Copy the data to their destination partitions and flush when the buffer is full.
            for (int i = 0; i < partitionBuffers.length; i++) {
                partitionBuffers[i].appendData(decodedBlocks, pageRetainedSizeInBytes, fixedWidthRowSize, variableWidthChannels, outputBuffer);
            }
...
            for (int i = 0; i < channelCount; i++) {
                blockEncodingBuffers[i].setupDecodedBlocksAndPositions(decodedBlocks[i], positions, positionCount, capacity, pageRetainedSizeInBytes);
            }
```

- Calculate block's fraction of the input page size in setupDecodedBlocksAndPositions and pass it to setupDecodedBlockAndMapPositions (decodedBlockPageSizeFraction below needs a better name, but I can't thing of one just yet):

```
    @Override
    public void setupDecodedBlocksAndPositions(DecodedBlockNode decodedBlockNode, int[] positions, int positionCount, int partitionBufferCapacity, long pageRetainedSizeInBytes)
    {
        requireNonNull(decodedBlockNode, "decodedBlockNode is null");
        requireNonNull(positions, "positions is null");

        this.positions = positions;
        this.positionCount = positionCount;
        this.positionsOffset = 0;
        this.positionsMapped = false;

        double decodedBlockPageSizeFraction = (decodedBlockNode.getRetainedSizeInBytes() - decodedBlockNode.getChildrenRetainedSizeInBytes()) / ((double) pageRetainedSizeInBytes);

        setupDecodedBlockAndMapPositions(decodedBlockNode, partitionBufferCapacity, decodedBlockPageSizeFraction);
    }
```

- Replace calls to DecodedBlock.getSizePortionInPage() with decodedBlockPageSizeFraction

yingsu00 commented:
> @yingsu00 Ying, thanks for explaining. I'm wondering if this calculation is correct when some blocks are dictionary encoded. Since repartitioning flattens these blocks, the fraction of the original page size for these blocks is less than the fraction of flattened data in the output page. Is this a problem? Would it help to account for that when computing the fractions?

@mbasmanova Hi Masha, yes the calculation for DictionaryBlock may be inaccurate. I thought of this before but couldn't find a good way to get an accurate estimation for the block after unpacking. One possible estimation is `positionCount * averageRowSize of dictionary`. What do you think about it? Just to mention that this won't cause functional issues but may cause the buffers to be grown slower.

And thanks for the suggestion of moving the logic into `setupDecodedBlocksAndPositions`. Actually I tried to do as much arithmetic calculation as possible just once for each incoming page. By moving the logic inside of the BlockEncodingBuffer, we're repeating the calculation of minus and division for all partitions/BlockEncodingBuffers which is unnecessary. If we do the minus and division in DecodedBlock, then we just need to do a multiplication in each partitions/BlockEncodingBuffers. Originally I was thinking to calculate the buffer size in DecodedBlock but they are orthogonal and may not be a good idea. Are you strongly against the current way? If yes I can do a JMH run to see if it has any regressions.

yingsu00 commented:
> > @yingsu00 Ying, thanks for explaining. I'm wondering if this calculation is correct when some blocks are dictionary encoded. Since repartitioning flattens these blocks, the fraction of the original page size for these blocks is less than the fraction of flattened data in the output page. Is this a problem? Would it help to account for that when computing the fractions?
> 
> @mbasmanova Hi Masha, yes the calculation for DictionaryBlock may be inaccurate. I thought of this before but couldn't find a good way to get an accurate estimation for the block after unpacking. One possible estimation is `positionCount * averageRowSize of dictionary`. What do you think about it? Just to mention that this won't cause functional issues but may cause the buffers to be grown slower.
> 
> And thanks for the suggestion of moving the logic into `setupDecodedBlocksAndPositions`. Actually I tried to do as much arithmetic calculation as possible just once for each incoming page. By moving the logic inside of the BlockEncodingBuffer, we're repeating the calculation of minus and division for all partitions/BlockEncodingBuffers which is unnecessary. If we do the minus and division in DecodedBlock, then we just need to do a multiplication in each partitions/BlockEncodingBuffers. Originally I was thinking to calculate the buffer size in DecodedBlock but they are orthogonal and may not be a good idea. Are you strongly against the current way? If yes I can do a JMH run to see if it has any regressions.

@mbasmanova  Hi Masha, I changed the code as you suggested and just pushed the PR. No regression was observed. Thank you for the detailed write-out! Will you take another look?

mbasmanova commented:
@yingsu00 I think `Introduce sizePortionInPage in DecodedBlockNode` commit needs to be removed.

> yes the calculation for DictionaryBlock may be inaccurate. I thought of this before but couldn't find a good way to get an accurate estimation for the block after unpacking. One possible estimation is positionCount * averageRowSize of dictionary. What do you think about it? 

Sounds reasonable.

yingsu00 commented:
> positionCount * averageRowSize
@mbasmanova Hi Masha, I removed `Introduce sizePortionInPage in DecodedBlockNode` commit and added a new commit `Use inflated size for dictionary and RLE blocks in DecodedBlockNode` to show you the change for DictionaryBlock and RunLengthEncodedBlock retained size estimation. But please also note that for Block views, this estimation is still inaccurate. The retained size is the size for whole arrays but the positionCount for the raw ids is not accessible. If we want to solve this problem, we could maintain both sizeInBytes and retainedSizeInBytes for DecodedBlockNode, and use sizeInBytes to estimate the buffer growth limit since it reflects only the size of the region for the views. What do you think?

mbasmanova commented:
@yingsu00 

>  But please also note that for Block views, this estimation is still inaccurate.

What is "Block view"?

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java:
```diff
@@ -479,13 +480,15 @@ public void partitionPage(Page page)
             // Copy the data to their destination partitions and flush when the buffer is full.
             for (int i = 0; i < partitionBuffers.length; i++) {
-                partitionBuffers[i].appendData(decodedBlocks, page.getRetainedSizeInBytes(), fixedWidthRowSize, variableWidthChannels, outputBuffer);
+                partitionBuffers[i].appendData(decodedBlocks, decodedBlocksRetainedSize, fixedWidthRowSize, variableWidthChannels, outputBuffer);
```
@yingsu00 I think moving away from page.getRetainedSizeInBytes() and using estimated flattened size is the right direction. It allows for much cleaner design. The DecodedBlockNode still needs the original value for the retainedSizeInBytes as it is used in memory tracking. Hence, I suggest to introduce a new variable to store the estimate of the flattened size of the block.

yingsu00 commented:
> @yingsu00
> 
> > But please also note that for Block views, this estimation is still inaccurate.
> 
> What is "Block view"?

By "Block view" I meant the block with non-zero offsetBase.

yingsu00 commented:
@mbasmanova Hi Masha, I found the Block interface has getLogicalSizeInBytes() and it can be used to estimate the serialized sizes. For DictionaryBlock and RunLengthEncodedBlock, it is the size after the block is expanded. For other blocks is the sizeInBytes which just account for the region that is going to be serialized. I think it's pretty accurate for our purpose. I updated the PR with two new commits
```
Introduce estimatedSerializedSizeInBytes in ColumnarXXX
Introduce estimatedSerializedSizeInBytes in DecodedBlockNode
```
I also modified `Enforce buffer size limits for BlockEncodingBuffer` to use the newly introduced `estimatedSerializedSizeInBytes`, and removed `Introduce sizePortionInPage in DecodedBlockNode`. The tests have already passed. Would appreciate your review again. Thank you!
